source/common/common/hash.h

```diff
@@ -26,6 +27,42 @@ class HashUtil {
     return XXH64(input.data(), input.size(), seed);
   }
 
+  /**
+   * Return 64-bit hash from deterministically serializing a value in
+   * an endian-independent way and hashing it with the xxHash algorithm.
+   *
+   * Floats and doubles are just done as endian-adjusted raw bytes the same
+   * as integers; this is on par with what protobuf serialization does.
+   * src/google/protobuf/generated_message_util.cc literally just treats
+   * a float the same as int32 for serialization purposes.
+   *
+   * enums are excluded because they're not needed; if they were to be
+   * supported later they should have a separate function because enum
+   * sizes are implementation-specific.
+   *
+   * bools have an inlined specialization below because they too have
+   * implementation-specific sizes.
+   *
+   * @param input supplies the value to hash.
+   * @param seed supplies the hash seed which defaults to 0.
+   * See https://github.com/Cyan4973/xxHash for details.
+   */
+  template <
+      typename ValueType,
+      std::enable_if_t<std::is_scalar_v<ValueType> && !std::is_enum_v<ValueType>, bool> = true>
+  static uint64_t xxHash64Value(ValueType input, uint64_t seed = 0) {
+#if defined(ABSL_IS_LITTLE_ENDIAN)
+    return XXH64(reinterpret_cast<const char*>(&input), sizeof(input), seed);
+#else
+    char buf[sizeof(input)];
```

Comment:
This is going to punish big-endian platforms, but it is probably not very relevant to Envoy. Unless someone will decide to build it as big-endian for Arm or PowerPC.

Out of curiosity, what context did this come up?

Reply:
Compared to the total human-readable stringification of the protobuf for hashing it will still be an *improvement* for big-endian in https://github.com/envoyproxy/envoy/pull/30761, just not as big of an improvement as it is for little-endian.

There are two main contexts where this hashing will be used, both targets of that PR - one is the cache filter (hence that contrived example where endianness might potentially make a difference) and the other is something during startup and config changes that hashes the config protobuf, which was the inspiration for making the change at all (because about 50% of the total time there is spent hashing protobufs right now, and with some of our configs this can be over 60 seconds).

If it became an issue, endianness-ignoring could be made an alternative option, but I don't think it needs to be done now - an incremental improvement to performance now is better than perfection never. (This performance cost has been a thing for years without being addressed.)

Reply:
Aha, thought of a better way to experiment it - with an external function with no definition. It does indeed boil the loop down to a single bswap instruction. https://godbolt.org/z/jTW8qMT63